File: engine-python/engine/rules.py
"""Capture legality, terrain effects, and win detection for Jungle."""

from __future__ import annotations

from config import (
    ROWS, TERRAIN, TERRAIN_RIVER,
    DEN_BLACK, DEN_BLUE, TRAPS_BLACK, TRAPS_BLUE,
    IS_RIVER, TRAP_ZEROES,
)
from engine.pieces import Animal, Color, piece_id_color, piece_id_animal, piece_id_rank


def _opponent_traps(color: Color) -> set[tuple[int, int]]:
    """Traps that reduce rank for the given color (i.e. the *opponent's* traps)."""
    return TRAPS_BLACK if color == Color.BLUE else TRAPS_BLUE


def effective_rank(pid: int, col: int, row: int) -> int:
    """Return the *defensive* effective rank of a piece at (col, row).

    A piece inside the *opponent's* traps has effective rank 0. This applies to
    defence only — see :func:`can_capture_sq`, which never applies it to the
    attacker.
    """
    color = piece_id_color(pid)
    if (col, row) in _opponent_traps(color):
        return 0
    return piece_id_rank(pid)


def can_capture_sq(attacker_pid: int, defender_pid: int,
                   atk_sq: int, def_sq: int, board) -> bool:
    """Flat-square capture legality — the engine's hottest predicate.

    The four checks below run in this order, and the order is load-bearing.

    1. **Enemies.** No friendly fire.
    2. **The water boundary**, absolute and symmetric: a piece in the river and a
       piece on land can never take one another, whichever way round. This is
       what makes a Rat in the river invulnerable to the bank, and what stops a
       Rat in the river taking an Elephant standing on it. It is checked before
       anything else because it outranks every other consideration.
    3. **A trapped defender** — one standing in the *attacker's* traps — has
       effective rank 0 and falls to any adjacent enemy, Elephant included, which
       is why this precedes the Rat/Elephant exception rather than following it.
    4. **Rank**, with the Rat/Elephant exception: Rat takes Elephant, Elephant
       does not take Rat, otherwise attacker rank >= defender rank.

    **The trap weakens a piece for defence only.** The attacker always fights at
    its real rank, so a piece standing in the enemy's traps is vulnerable but not
    disarmed. Applying the rank-0 effect to the attacker as well silently makes a
    class of legal captures illegal near both dens — it did here until the Rust
    engine's golden corpus disagreed on 14 positions in 10,000.

    Works on plain int ranks (Animal is an IntEnum: Rat == 1, Elephant == 8)
    and flat config tables (IS_RIVER, TRAP_ZEROES).
    """
    atk_blue = attacker_pid > 0

    # 1. Must be enemies
    if atk_blue == (defender_pid > 0):
        return False

    # 2. The water boundary, for every piece and not just the Rat
    if IS_RIVER[atk_sq] != IS_RIVER[def_sq]:
        return False

    # 3. A defender in the attacker's traps falls to anything
    if TRAP_ZEROES[def_sq] == (1 if defender_pid > 0 else 2):
        return True

    atk_rank = attacker_pid if atk_blue else -attacker_pid
    def_rank = defender_pid if defender_pid > 0 else -defender_pid

    # 4. Rank, with the Rat/Elephant exception. The attacker keeps its real rank.
    if atk_rank == 1 and def_rank == 8:
        return True
    if atk_rank == 8 and def_rank == 1:
        return False
    return atk_rank >= def_rank


def can_capture(attacker_pid: int, defender_pid: int,
                atk_col: int, atk_row: int,
                def_col: int, def_row: int,
                board) -> bool:
    """(col, row) wrapper around :func:`can_capture_sq` (public rules API)."""
    return can_capture_sq(attacker_pid, defender_pid,
                          atk_col * ROWS + atk_row, def_col * ROWS + def_row,
                          board)


def is_jump_blocked(fc: int, fr: int, tc: int, tr: int, board) -> bool:
    """Return True if a river-jump from (fc,fr) to (tc,tr) is blocked by a rat in the river.

    The jump must be a straight line crossing one of the two river blocks.
    Any rat (either color) on a water square along the path blocks the jump.
    """
    if fc == tc:
        # Vertical jump
        step = 1 if tr > fr else -1
        r = fr + step
        while r != tr:
            if TERRAIN[fc][r] == TERRAIN_RIVER:
                pid = board.get(fc, r)
                if pid != 0 and piece_id_animal(pid) == Animal.RAT:
                    return True
            r += step
    elif fr == tr:
        # Horizontal jump
        step = 1 if tc > fc else -1
        c = fc + step
        while c != tc:
            if TERRAIN[c][fr] == TERRAIN_RIVER:
                pid = board.get(c, fr)
                if pid != 0 and piece_id_animal(pid) == Animal.RAT:
                    return True
            c += step
    return False


class WinResult:
    __slots__ = ("winner",)

    def __init__(self, winner: Color) -> None:
        self.winner = winner


def check_win(board, last_move, current_turn: Color) -> WinResult | None:
    """Check if the game has ended after last_move was played.

    Win conditions:
    1. A piece entered the opponent's den.
    2. One side has no pieces remaining.
    3. The player whose turn it is has no legal moves (stalemate → they lose).

    'current_turn' is the side about to move (i.e. the side that did NOT just move).
    """
    if last_move is not None:
        # Condition 1: den entry
        mover_pid = board.get(last_move.tc, last_move.tr)
        if mover_pid != 0:
            mover_color = piece_id_color(mover_pid)
            opponent_den = DEN_BLUE if mover_color == Color.BLACK else DEN_BLACK
            if (last_move.tc, last_move.tr) == opponent_den:
                return WinResult(mover_color)

    # Condition 2: capture all
    for color in (Color.BLUE, Color.BLACK):
        if board.alive_count(color) == 0:
            opponent = Color.BLACK if color == Color.BLUE else Color.BLUE
            return WinResult(opponent)

    return None

File: rust/crates/jungle-core/src/rules.rs
//! Capture legality and effective rank.
//!
//! The order of the checks in [`can_capture`] is load-bearing and is the one
//! thing here that must not be "simplified". The project's one recorded rules
//! defect was the Elephant-cannot-take-Rat rejection sitting *before* the trap
//! override, which made a Rat in its own trap untouchable by an Elephant that it
//! could take back. Water boundary, then trapped defender, then rank.

use crate::bitboard::{bb, is_river, weakening_traps};
use crate::types::{animal, Piece, Square};

/// A piece standing in the *enemy's* home traps fights at rank 0.
///
/// This applies to the **defender only**. An attacker standing in enemy traps
/// keeps its real rank: a trapped piece is vulnerable, not disarmed.
#[inline(always)]
pub fn effective_rank(piece: Piece, s: Square) -> u8 {
    if weakening_traps(piece.color().index()) & bb(s) != 0 {
        0
    } else {
        piece.rank()
    }
}

/// Can `attacker` on `atk_sq` take `defender` on `def_sq`?
#[inline]
pub fn can_capture(attacker: Piece, defender: Piece, atk_sq: Square, def_sq: Square) -> bool {
    // (1) Must be enemies.
    if attacker.color() == defender.color() {
        return false;
    }

    // (2) No capture across the water/land boundary, in either direction. This
    //     is what makes a Rat in the river invulnerable to land pieces, and what
    //     stops a river Rat from taking an Elephant on the bank.
    if is_river(atk_sq) != is_river(def_sq) {
        return false;
    }

    // (3) A defender in the attacker's home traps has rank 0 and falls to
    //     anything — including an Elephant taking a Rat. Traps are never river
    //     squares, so this cannot bypass (2).
    if effective_rank(defender, def_sq) == 0 {
        return true;
    }

    // (4) Rank comparison, with the Rat/Elephant exception. The attacker always
    //     uses its real rank.
    let a = attacker.rank();
    let d = defender.rank();
    if a == animal::RAT && d == animal::ELEPHANT {
        return true;
    }
    if a == animal::ELEPHANT && d == animal::RAT {
        return false;
    }
    a >= d
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::bitboard::{DEN_BLACK, TRAPS_BLACK};
    use crate::types::{sq, Color};

    fn blue(rank: u8) -> Piece {
        Piece::new(Color::Blue, rank)
    }
    fn black(rank: u8) -> Piece {
        Piece::new(Color::Black, rank)
    }

    #[test]
    fn higher_or_equal_rank_captures() {
        let a = sq(0, 4);
        let d = sq(0, 5);
        assert!(can_capture(blue(5), black(4), a, d));
        assert!(can_capture(blue(4), black(4), a, d), "equal ranks trade");
        assert!(!can_capture(blue(3), black(4), a, d));
    }

    #[test]
    fn rat_beats_elephant_but_not_the_reverse() {
        let a = sq(0, 4);
        let d = sq(0, 5);
        assert!(can_capture(blue(animal::RAT), black(animal::ELEPHANT), a, d));
        assert!(!can_capture(blue(animal::ELEPHANT), black(animal::RAT), a, d));
    }

    #[test]
    fn captures_never_cross_the_water_boundary() {
        let water = sq(1, 4);
        let land = sq(0, 4);
        assert!(!can_capture(blue(animal::RAT), black(animal::ELEPHANT), water, land));
        assert!(!can_capture(blue(animal::ELEPHANT), black(animal::RAT), land, water));
        // Rat vs Rat on the same terrain is fine.
        assert!(can_capture(blue(animal::RAT), black(animal::RAT), water, sq(2, 4)));
    }

    #[test]
    fn a_trapped_defender_falls_to_anything() {
        // A Blue piece standing in Black's home traps is the weakened one.
        for trap in [sq(2, 0), sq(4, 0), sq(3, 1)] {
            assert!(TRAPS_BLACK & bb(trap) != 0);
            for victim in 1..=8u8 {
                for attacker in 1..=8u8 {
                    assert!(
                        can_capture(black(attacker), blue(victim), DEN_BLACK, trap),
                        "black {attacker} should take trapped blue {victim}"
                    );
                }
            }
        }
    }

    #[test]
    fn a_trapped_attacker_keeps_its_rank() {
        // Blue Rat inside Black's trap still cannot be taken by an Elephant that
        // is itself standing in Blue's trap -- the trap weakens defence only.
        let blue_trap = sq(3, 7);
        let adjacent = sq(3, 6);
        assert!(!can_capture(
            black(animal::ELEPHANT),
            blue(animal::RAT),
            blue_trap,
            adjacent
        ));
        // ... and it still strikes at full power.
        assert!(can_capture(
            black(animal::ELEPHANT),
            blue(animal::LION),
            blue_trap,
            adjacent
        ));
    }

    #[test]
    fn own_colour_never_captures() {
        assert!(!can_capture(blue(8), blue(1), sq(0, 4), sq(0, 5)));
    }
}
